site/en/models.md

```diff
@@ -12,118 +12,251 @@ its own internal format(s) for models. Some support central storage of models
 
 To use a specific model with OCR-D's calamari wrapper
 [ocrd_calamari](https://github.com/OCR-D/ocrd_calamari) and more specifically,
-the `ocrd-calamari-recognize` processor, use the `checkpoint` parameter:
+the `ocrd-calamari-recognize` processor, use the `checkpoint_dir` parameter:
 
 ```sh
-ocrd-calamari-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-CALA -p '{"checkpoint": "/path/to/model/*.ckpt.json"}'
+# To use the "default" model, i.e. the one trained on GT4HistOCR by QURATOR
+ocrd-calamari-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-CALA
+# To use your own trained model
+ocrd-calamari-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-CALA -P checkpoint_dir /path/to/modeldir
+# or, to be able to control which checkpoints to use:
+ocrd-calamari-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-CALA -P checkpoint '/path/to/modeldir/*.ckpt.json'
 ```
 
+## Tesseract / ocrd_tesserocr
+
+Tesseract models are single files with a `.traineddata` extension.
+
+Since tesseract only supports model lookup in a single directory, models should
+only be stored in a single location. If the default location (`virtualenv`) is
+not the place you want to use for tesseract models, consider [changing the default location
+in the OCR-D config file](#changing-the-default-resource-directory).
+
+**NOTE:** For reasons of effiency and to avoid duplicate models, all `ocrd-tesserocr-*` processors
+reuse the resource directory for `ocrd-tesserocr-recognize`.
+
+If the `TESSDATA_PREFIX` environemnt variable is set when any of the tesseract processors
+are called, it will be the location to look for resources instead of the default.
+
+OCR-D's Tesseract wrapper,
+[ocrd_tesserocr](https://github.com/OCR-D/ocrd_tesserocr) and more
+specifically, the `ocrd-tesserocr-recognize` processor, expects the name of the
+model(s) to be provided as the `model` parameter. Multiple models can be
+combined by concatenating with `+` (which generally improves accuracy but always slows processing):
+
+```sh
+# Use the deu and frk models
+ocrd-tesserocr-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-TESS -P model 'deut+frk'
+# Use the Fraktur model
+ocrd-tesserocr-recognize -I OCR-D-SEG-LINE -O OCR-D-OCR-TESS -P Fraktur
+```
+
+# Models and docker
+
+We recommend a two-step process to make models available in Docker. First
+download all the models that you want to use on the host system. When running
+the docker container, mount that local directory into the container alongside
+the data you want to process.
+
+Download the models to `$HOME/.local/share/ocrd-resources`:
+
+```sh
+ocrd resmgr download --location data ocrd-tesserocr-recognize eng.traineddata
+ocrd resmgr download --location data ocrd-calamari-recognize default
+# ...
+```
+
+Run the `ocrd_all` Docker container:
+
+```sh
+docker run --user $(id -u) --workdir /data \
+  --volume $PWD:/data \
+  --volume $HOME/.local/cache/ocrd-resources:/ocrd-resources \
+  ocrd_all ocrd-tesserocr-recognize -I IN -O OUT -P model eng
+```
```

Comment:
```suggestion
We recommend keeping all downloaded resources in a persistent host directory, 
separate of the `ocrd/all` Docker container and data directory, and mounting that
resource directory into a specific path in the container alongside the data directory.
The host resource directory can be empty initially. Each time you run the Docker container,
your processors will access the host directory to resolve resources, and you can download
additional models into that location using `ocrd resmgr`.

The following will assume (without loss of generality) that your host-side data path is under `./data`, and the host-side resource path is under `./models`:

- To download models to `/usr/share/ocrd-resources` in Docker and `./models` in the host FS, use `virtualenv` as storage location:

        docker run --user $(id -u) --workdir /data \
          --volume $PWD/data:/data \
          --volume $PWD/models:/usr/share/ocrd-resources \
        ocrd/all \
        ocrd resmgr download --location virtualenv ocrd-tesserocr-recognize eng.traineddata\; \
        ocrd resmgr download --location virtualenv ocrd-calamari-recognize default\; \
        ...

- To run processors, as usual do:

        docker run --user $(id -u) --workdir /data \
          --volume $PWD/data:/data \
          --volume $PWD/models:/usr/share/ocrd-resources \
          ocrd/all ocrd-tesserocr-recognize -I IN -O OUT -P model eng

```

Reply:
And IIRC for that very reason we also wanted to make `--location virtualenv` the _default_ in the Docker build...

Reply:
> > And IIRC for that very reason we also wanted to make `--location virtualenv` the _default_ in the Docker build...
> 
> This works on the top-level venv but what about the sub-venvs? They have their own `$VIRTUAL_ENV` and hence the `ocrd resmgr` in the sub-venvs will not see the model installed to top-level `$VIRTUAL_ENV/share/ocrd-resources` :( Correct me if I'm wrong but I think we do need either XDG or re-introduce an absolute path like `/usr/share/local/ocrd-resources` ...

Uh-oh, yes, you are right! But it's not `ocrd resmgr` which is the problem – since we would only need that at the top-level venv – but the processors' `resource_manager` search path. So `virtualenv` is out of question not only for Docker, but also for native ocrd_all deployments, right? Therefore we must always use one of the XDG paths, except maybe for users who install modules individually. But the latter can still live with XDG I guess, so I recommend removing `virtualenv` **entirely** from the implementation and spec.

Do you want me to update the above suggestion in terms of XDG `data`?

Reply:
> So virtualenv is out of question not only for Docker, but also for native ocrd_all deployments, right?

Afraid so :( I'll adapt the spec and implementation to default to `data` for consistency.

I am also not happy about `$HOME` being `/` or `/root` depending on the `docker run` call. Reintroduce /usr/local/share (shorthand `system`)?

> Do you want me to update the above suggestion in terms of XDG data?

Yes, please, thanks.

Reply:
> > So virtualenv is out of question not only for Docker, but also for native ocrd_all deployments, right?
> 
> Afraid so :( I'll adapt the spec and implementation to default to `data` for consistency.

You mean `and default to`, right? (We should rid of `virtualenv` as a possibility itself.)

> I am also not happy about `$HOME` being `/` or `/root` depending on the `docker run` call. Reintroduce /usr/local/share (shorthand `system`)?

Yes, we don't want to get into that mess. We could from now on set `ENV HOME=/` in our Dockerfiles (ocrd/all as well as ocrd/tesserocr etc), so `XDG_DATA_HOME` would be `/.local/share`, or even set `ENV XDG_DATA_HOME=/usr/local/share` and `ENV XDG_CONFIG_HOME=/usr/local/etc` and so on. 

Regarding absolute paths like `/usr/local/share/ocrd-resources` – agreed, but only if configurable (not fixed). So if `XDG_CONFIG_HOME` can be relied upon (even in Docker, somehow), then core can always find its `ocrd/config.yml` (the same for top-level and sub venvs), look up the default `resouce_location` (from the enum `data` / `cache` / `config` or a string for the absolute path), and be done. 

That one assumption (`XDG_CONFIG_HOME`) _must_ be satisfiable – after all, what's the point of configuring paths if you cannot even reliably find the path to that config?

Reply:
In light of the recent discussion, with `virtualenv` disqualifying itself due to our sub-venv mechanics, and the need to have at least one stable entry point for configuration, I wonder what's the point of keeping the many-valued `resource_location` logic and a config file for it at all. The deployment just needs to make sure `XDG_DATA_HOME` is set sensibly (for `$XDG_DATA_HOME/ocrd-resources`), everything else is unnecessary complexity.

Reply:
BTW: considering that https://github.com/OCR-D/ocrd_all/pull/234 simply delegates to `ocrd resmgr`, we could even explain to just do

    docker run --user $(id -u) --volume $PWD/models:/usr/local/share/ocrd-resources \
        ocrd/all make -C /build install-models

Do you want me to include even that in the above suggestion?

Reply:
> BTW: considering that [OCR-D/ocrd_all#234](https://github.com/OCR-D/ocrd_all/pull/234) simply delegates to `ocrd resmgr`, we could even explain to just do
> 
> ```
> docker run --user $(id -u) --volume $PWD/models:/usr/local/share/ocrd-resources \
>     ocrd/all make -C /build install-models
> ```
> 
> Do you want me to include even that in the above suggestion?

I love that this is possible but I do think that we should familiarize users with `ocrd resmgr`, whereas `make install-models` was just a stop-gap solution.

Reply:
> I love that this is possible but I do think that we should familiarize users with `ocrd resmgr`, whereas `make install-models` was just a stop-gap solution.

Okay how about a `ocrd resmgr download '*'` then?

In contrast to `install-models`, this leaves the problem of knowing which processors/modules are actually installed to the user (or to `ocrd resmgr`).

Reply:
> EDIT: What should happen if I `ocrd resmgr download ocrd-kraken-recognize '*'` although I don't have ocrd_kraken installed

It will install the models just fine. You could also do `ocrd resmgr download ocrd-foo /vmlinuz` and it would copy `/vmlinuz` to `$HOME/.local/share/ocrd-foo/vmlinuz`, no questions asked.

> In contrast to install-models, this leaves the problem of knowing which processors/modules are actually installed to the user (or to ocrd resmgr), though.

I also saw your proposal in ocrd_all but, while an elegant use of make, it artificially restricts what can be done. From experience, there probably are unexpected use cases, where that restriction would require users to hack around those rules.

> Okay how about a ocrd resmgr download '*' then?

What we could do is: Find all `ocrd-*` in `$PATH` and use that as the set of processors for which to download resources. Would that achieve what you want?

Reply:
> > What should happen if I `ocrd resmgr download ocrd-kraken-recognize '*'` although I don't have ocrd_kraken installed
> 
> It will install the models just fine. You could also do `ocrd resmgr download ocrd-foo /vmlinuz` and it would copy `/vmlinuz` to `$HOME/.local/share/ocrd-foo/vmlinuz`, no questions asked.

IMHO we should at least warn that there's no `ocrd-foo` here. (Maybe it was misspelled, or the user forgot to install it?)

> I also saw your proposal in ocrd_all but, while an elegant use of make, it artificially restricts what can be done. From experience, there probably are unexpected use cases, where that restriction would require users to hack around those rules.

If you are arguing against [encapsulation in ocrd_all](https://github.com/OCR-D/ocrd_all/pull/234#pullrequestreview-574250990), then I must respond that just as `OCRD_MODULES` can be changed freely to extend or reduce the set of modules to install, here `OCRD_MODULES` would be used for the set of models to install – it's what should be expected for `make install-models` per analogy.

But if all you wanted to say was that `ocrd resmgr download '*'` should not suffer any restrictions, then I tend to agree with you. However, remember we still want to aggregate the `resource_list.yml` during ocrd_all's `make all` in the long term. So naturally the list of available models would derive from the list of installed modules. The connection is already there! If the user installs other modules manually, fine, they will have only unregistered resources. Still, `*` should give only the registered models (i.e. from modules included in `make all`).

Reply:
> IMHO we should at least warn that there's no ocrd-foo here. (Maybe it was misspelled, or the user forgot to install it?)

Agreed, 22fb2c62. 

> If you are arguing against encapsulation in ocrd_all, then I must respond that just as OCRD_MODULES can be changed freely to extend or reduce the set of modules to install, here OCRD_MODULES would be used for the set of models to install – it's what should be expected for make install-models per analogy.

Point taken, lets continue discussing at https://github.com/OCR-D/ocrd_all/pull/234#pullrequestreview-574250990

Reply:
Replaced the section with your proposed text in f76c0a7b

BTW `ocrd resmgr download '*'` is now implemented in OCR-D/core#559.

Reply:
> BTW `ocrd resmgr download '*'` is now implemented in [OCR-D/core#559](https://github.com/OCR-D/core/pull/559).

Fantastic!